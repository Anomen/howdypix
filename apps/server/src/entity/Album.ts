import { Column, ViewEntity } from "typeorm";

@ViewEntity({
  expression: `
SELECT
   P.DIR,
   P.SOURCE,
   P.PARENTDIR,
   COUNT(DISTINCT P2.ID) AS NBPHOTOS,
   COUNT(DISTINCT P3.DIR) AS NBALBUMS,
   (
      SELECT
         FILE 
      FROM
         PHOTO P4 
      WHERE
         P4.DIR = P.DIR 
         AND P4.SOURCE = P.SOURCE LIMIT 0,
         1
   )
   AS PREVIEW 
FROM
   PHOTO P 
   LEFT JOIN
      PHOTO P2 
      ON (P2.DIR = P.DIR 
      AND P2.SOURCE = P.SOURCE) 
   LEFT JOIN
      PHOTO P3 
      ON (P3.PARENTDIR = P.DIR 
      AND P3.SOURCE = P.SOURCE) 
GROUP BY
   P.DIR,
   P.SOURCE,
   P.PARENTDIR;
    `,
})
export class Album {
  @Column("text")
  dir: string;

  @Column("text")
  source: string;

  @Column("text")
  parentDir: string;

  @Column("text")
  preview: string;

  @Column("int")
  nbPhotos: number;

  @Column("int")
  nbAlbums: number;
}
