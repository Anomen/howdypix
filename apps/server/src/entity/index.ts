export * from "./Source";
export * from "./Album";
export * from "./Photo";
export * from "./Search";
export * from "./SearchResult";
