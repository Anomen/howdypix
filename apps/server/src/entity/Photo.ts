import {
  Column,
  Entity,
  ManyToOne,
  OneToMany,
  PrimaryGeneratedColumn,
  Unique,
} from "typeorm";
import { SearchResult } from "./SearchResult";
import { Album } from "./Album";

export enum PHOTO_STATUS {
  NOT_PROCESSED = "not_processed",
  PROCESSED = "processed",
}

@Entity()
@Unique(["album", "file"])
export class Photo {
  @PrimaryGeneratedColumn()
  id: number;

  @Column("text")
  status: PHOTO_STATUS;

  @Column("int")
  inode: number;

  @Column("int")
  mtime: number;

  @Column("int")
  ctime: number;

  @Column("int")
  birthtime: number;

  @Column("int")
  size: number;

  @Column("text", { nullable: true })
  make: string;

  @Column("text", { nullable: true })
  model: string;

  @Column("int", { nullable: true })
  ISO: number;

  @Column("int", { nullable: true })
  shutter: number;

  @Column("int", { nullable: true })
  processedShutter: number;

  @Column("int", { nullable: true })
  aperture: number;

  @Column("int", { nullable: true })
  processedAperture: number;

  @Column("int", { nullable: true })
  createDate: number;

  // TODO to remove
  @Column("text")
  source: string;

  // TODO to remove
  @Column("text")
  parentDir: string;

  // TODO to remove
  @Column("text")
  dir: string;

  @ManyToOne(() => Album, (album) => album.photos)
  public album!: Album;

  @Column("text")
  file: string;

  @OneToMany(() => SearchResult, (searchResult) => searchResult.photo)
  public searchResults!: SearchResult[];
}
