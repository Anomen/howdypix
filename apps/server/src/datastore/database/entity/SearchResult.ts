import {
  Column,
  Entity,
  PrimaryGeneratedColumn,
  ManyToOne,
  CreateDateColumn,
  UpdateDateColumn,
} from "typeorm";
import { Search } from "./Search";
import { Photo } from "./Photo";

@Entity()
export class SearchResult {
  @PrimaryGeneratedColumn()
  id: number;

  @CreateDateColumn()
  createdAt: Date;

  @UpdateDateColumn()
  updatedAt: Date;

  @Column("int")
  searchId!: number;

  @Column("int")
  photoId!: number;

  @Column("int")
  public order!: number;

  @ManyToOne(() => Search, (search) => search.searchResults)
  public search!: Search;

  @ManyToOne(() => Photo, (photo) => photo.searchResults)
  public photo!: Photo;
}
