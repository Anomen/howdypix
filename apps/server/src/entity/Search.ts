import { Column, Entity, PrimaryGeneratedColumn, OneToMany } from "typeorm";
import { SearchResult } from "./SearchResult";

@Entity()
export class Search {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ type: "text", nullable: true })
  orderBy: string;

  @Column({ type: "text", nullable: true })
  filterBy: string;

  @Column({ type: "text", nullable: true })
  album: string;

  @Column({ type: "text", nullable: true })
  source: string;

  @OneToMany(
    () => SearchResult,
    searchResult => searchResult.search
  )
  public searchResults!: SearchResult[];
}
