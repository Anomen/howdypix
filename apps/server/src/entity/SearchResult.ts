import { Column, Entity, PrimaryGeneratedColumn, ManyToOne } from "typeorm";
import { Search } from "./Search";
import { Photo } from "./Photo";

@Entity()
export class SearchResult {
  @PrimaryGeneratedColumn()
  id: number;

  @Column("int")
  searchId!: number;

  @Column("int")
  photoId!: number;

  @Column("int")
  public order!: number;

  @ManyToOne(
    () => Search,
    search => search.searchResults
  )
  public search!: Search;

  @ManyToOne(
    () => Photo,
    photo => photo.searchResults
  )
  public photo!: Photo;
}
