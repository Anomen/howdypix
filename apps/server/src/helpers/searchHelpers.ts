import { NexusGenArgTypes } from "@howdypix/graphql-schema/schema.d";
import {
  FindOneOptions,
  Repository,
  Connection,
  Between,
  FindOperator,
} from "typeorm";
import { sortJsonStringify } from "@howdypix/utils";
import { Photo as EntityPhoto } from "../entity/Photo";
import { SearchResult as EntitySearchResult } from "../entity/SearchResult";
import { Search as EntitySearch } from "../entity/Search";
import { filterByMake, filterByModel } from "../lib/filters";

export const searchHelpers = {
  getOrderBy: (
    order: NexusGenArgTypes["Query"]["getSearch"]["orderBy"]
  ): FindOneOptions<EntityPhoto>["order"] => {
    switch (order) {
      case "DATE_ASC":
        return { birthtime: "ASC" };
      case "DATE_DESC":
        return { birthtime: "DESC" };
      case "NAME_ASC":
        return { file: "ASC" };
      case "NAME_DESC":
        return { file: "DESC" };
      default:
        return { birthtime: "DESC" };
    }
  },

  findSavedSearch: async (
    searchRepository: Repository<EntitySearch>,
    args: NexusGenArgTypes["Query"]["getSearch"]
  ): Promise<EntitySearch | undefined> =>
    searchRepository.findOne({
      where: {
        orderBy: args.orderBy ?? null,
        filterBy: args.filterBy ? sortJsonStringify(args.filterBy) : null,
        album: args.album ?? "",
        source: args.source ?? "",
      },
    }),

  saveNewSearch: async (
    searchRepository: Repository<EntitySearch>,
    args: NexusGenArgTypes["Query"]["getSearch"]
  ): Promise<EntitySearch> => {
    const newSearch = new EntitySearch();
    newSearch.source = args.source ?? "";
    newSearch.album = args.album ?? "";

    if (args.orderBy) {
      newSearch.orderBy = args.orderBy;
    }

    if (args.filterBy) {
      newSearch.filterBy = sortJsonStringify(args.filterBy);
    }

    return searchRepository.save(newSearch);
  },

  doSearch: async (
    photoRepository: Repository<EntityPhoto>,
    args: NexusGenArgTypes["Query"]["getSearch"]
  ): Promise<EntityPhoto[]> => {
    // Define the where statement to filter the right photos
    const where: { dir?: string; source?: string } = {};
    where.dir = args.album ?? "";

    if (args.source) {
      where.source = args.source;
    }

    return photoRepository.find({
      where: {
        ...where,
        ...filterByMake(args.filterBy?.make).whereStatement,
        ...filterByModel(args.filterBy?.model).whereStatement,
      },
      order: searchHelpers.getOrderBy(args.orderBy),
    });
  },

  saveSearchResult: async (
    searchResultRepository: Repository<EntitySearchResult>,
    photos: EntityPhoto[],
    search: EntitySearch
  ): Promise<EntitySearchResult[]> =>
    Promise.all(
      photos.map(async (photo, key) => {
        const newSearchResult = new EntitySearchResult();
        newSearchResult.order = key;
        newSearchResult.photo = photo;
        newSearchResult.search = search;
        return searchResultRepository.save(newSearchResult);
      })
    ),

  fetchSearchResult: async (
    searchResultRepository: Repository<EntitySearchResult>,
    search: EntitySearch,
    start: number,
    limit: number | null
  ): Promise<EntitySearchResult[]> => {
    const where: { search: EntitySearch; order?: FindOperator<number> } = {
      search,
    };

    if (limit) {
      where.order = Between(start, start + limit);
    }

    return searchResultRepository.find({
      where,
      relations: ["photo"],
      order: { order: "ASC" },
    });
  },

  doSearchWithCache: async (
    connection: Connection,
    args: NexusGenArgTypes["Query"]["getSearch"],
    start = 0,
    limit: number | null = null
  ): Promise<EntitySearchResult[]> => {
    const searchRepository = connection.getRepository(EntitySearch);
    const photoRepository = connection.getRepository(EntityPhoto);
    const searchResultRepository = connection.getRepository(EntitySearchResult);

    const searchResults: EntitySearchResult[] = [];
    const search = await searchHelpers.findSavedSearch(searchRepository, args);

    if (!search) {
      // We create the new entry in the Search table
      const newSearch = await searchHelpers.saveNewSearch(
        searchRepository,
        args
      );

      // We search for the terms according to the criterias
      const photos = await searchHelpers.doSearch(photoRepository, args);

      (
        await searchHelpers.saveSearchResult(
          searchResultRepository,
          photos,
          newSearch
        )
      ).forEach((s) => {
        if (limit) {
          if (s.order >= start && s.order < start + limit) {
            searchResults.push(s);
          }
        } else {
          searchResults.push(s);
        }
      });
    } else {
      // Fetch the search results
      (
        await searchHelpers.fetchSearchResult(
          searchResultRepository,
          search,
          start,
          limit
        )
      ).forEach((s) => {
        searchResults.push(s);
      });
    }

    return searchResults;
  },
};
