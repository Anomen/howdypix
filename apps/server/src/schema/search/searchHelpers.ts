import { NexusGenArgTypes } from "@howdypix/graphql-schema/schema.d";
import { FindOneOptions, Repository } from "typeorm";
import { Photo as EntityPhoto } from "../../entity/Photo";
import { SearchResult as EntitySearchResult } from "../../entity/SearchResult";
import { Search as EntitySearch } from "../../entity/Search";

export const getOrderBy = (
  order: NexusGenArgTypes["Query"]["getSearch"]["orderBy"]
): FindOneOptions<EntityPhoto>["order"] => {
  switch (order) {
    case "DATE_ASC":
      return { birthtime: "ASC" };
    case "DATE_DESC":
      return { birthtime: "DESC" };
    case "NAME_ASC":
      return { file: "ASC" };
    case "NAME_DESC":
      return { file: "DESC" };
    default:
      return { birthtime: "DESC" };
  }
};

export const findSavedSearch = async (
  searchRepository: Repository<EntitySearch>,
  args: NexusGenArgTypes["Query"]["getSearch"]
): Promise<EntitySearch | undefined> =>
  searchRepository.findOne({
    where: {
      orderBy: args.orderBy,
      album: args.album ?? "",
      source: args.source ?? ""
    }
  });

export const saveNewSearch = async (
  searchRepository: Repository<EntitySearch>,
  args: NexusGenArgTypes["Query"]["getSearch"]
): Promise<EntitySearch> => {
  const newSearch = new EntitySearch();
  newSearch.source = args.source ?? "";
  newSearch.album = args.album ?? "";
  newSearch.orderBy = args.orderBy ?? "";

  return searchRepository.save(newSearch);
};

export const doSearch = async (
  photoRepository: Repository<EntityPhoto>,
  args: NexusGenArgTypes["Query"]["getSearch"]
): Promise<EntityPhoto[]> => {
  const where: { dir?: string; source?: string } = {};
  if (args.album) {
    where.dir = args.album;
  }
  if (args.source) {
    where.source = args.source;
  }

  return photoRepository.find({
    where,
    order: getOrderBy(args.orderBy)
  });
};

export const saveSearchResult = async (
  searchResultRepository: Repository<EntitySearchResult>,
  photos: EntityPhoto[],
  search: EntitySearch
): Promise<EntitySearchResult[]> => {
  return Promise.all(
    photos.map(async (photo, key) => {
      const newSearchResult = new EntitySearchResult();
      newSearchResult.order = key;
      newSearchResult.photo = photo;
      newSearchResult.search = search;
      return searchResultRepository.save(newSearchResult);
    })
  );
};

export const fetchSearchResult = async (
  searchResultRepository: Repository<EntitySearchResult>,
  search: EntitySearch
): Promise<EntitySearchResult[]> => {
  return searchResultRepository.find({
    where: {
      search
    },
    relations: ["photo"],
    order: { order: "ASC" }
  });
};
