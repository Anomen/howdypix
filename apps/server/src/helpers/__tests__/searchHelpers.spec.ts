import { Connection, Repository } from "typeorm";
import { searchHelpers } from "../searchHelpers";
import { Photo as EntityPhoto } from "../../entity/Photo";
import { SearchResult as EntitySearchResult } from "../../entity/SearchResult";
import { Search as EntitySearch } from "../../entity/Search";

const {
  doSearch,
  fetchSearchResult,
  findSavedSearch,
  saveNewSearch,
  saveSearchResult,
} = searchHelpers;

describe("searchHelpers", () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  test("findSavedSearch should make the right query", async () => {
    const findOne = jest.fn(() => ({ data: "yay" }));
    const repository = {
      findOne,
    };

    expect(
      await findSavedSearch(
        (repository as unknown) as Repository<EntitySearch>,
        {
          orderBy: "DATE_ASC",
          source: "source",
          album: "album",
        }
      )
    ).toMatchSnapshot();

    expect(findOne.mock.calls).toMatchSnapshot();
  });

  test("saveNewSearch should make the right query", async () => {
    const save = jest.fn(() => ({ data: "yay" }));
    const repository = {
      save,
    };

    expect(
      await saveNewSearch((repository as unknown) as Repository<EntitySearch>, {
        orderBy: "DATE_ASC",
        source: "source",
        album: "album",
      })
    ).toMatchSnapshot();

    expect(save.mock.calls).toMatchSnapshot();
  });

  test("doSearch should make the right query", async () => {
    const find = jest.fn(() => [{ data: "yay" }]);
    const repository = {
      find,
    };

    expect(
      await doSearch((repository as unknown) as Repository<EntityPhoto>, {
        orderBy: "DATE_ASC",
        source: "source",
        album: "album",
      })
    ).toMatchSnapshot();

    expect(find.mock.calls).toMatchSnapshot();
  });

  test("saveSearchResult should make the right query", async () => {
    const save = jest.fn(() => [{ data: "yay" }]);
    const repository = {
      save,
    };

    expect(
      await saveSearchResult(
        (repository as unknown) as Repository<EntitySearchResult>,
        // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
        // @ts-ignore
        [{ id: 123 }, { id: 456 }],
        { id: 890 }
      )
    ).toMatchSnapshot();

    expect(save.mock.calls).toMatchSnapshot();
  });

  test("fetchSearchResult should make the right query", async () => {
    const find = jest.fn(() => [{ data: "yay" }]);
    const repository = {
      find,
    };

    expect(
      await fetchSearchResult(
        (repository as unknown) as Repository<EntitySearchResult>,
        {
          id: 890,
          album: "album",
          source: "source",
          orderBy: "orderBy",
          searchResults: [],
          filterBy: "",
        },
        1,
        5
      )
    ).toMatchSnapshot();

    expect(find.mock.calls).toMatchSnapshot();
  });

  describe("doSearchWithCache", () => {
    beforeEach(() => {
      jest.resetAllMocks();

      // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
      // @ts-ignore
      searchHelpers.findSavedSearch = jest.fn();
      // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
      // @ts-ignore
      searchHelpers.saveSearchResult = jest.fn();
      // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
      // @ts-ignore
      searchHelpers.saveNewSearch = jest.fn();
      // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
      // @ts-ignore
      searchHelpers.doSearch = jest.fn();
      // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
      // @ts-ignore
      searchHelpers.fetchSearchResult = jest.fn();
    });

    const connection = ({
      getRepository: jest.fn(() => ({})),
    } as unknown) as Connection;

    describe("with not-saved results", () => {
      beforeEach(async () => {
        (searchHelpers.findSavedSearch as jest.Mock).mockReturnValue(null);
        (searchHelpers.saveSearchResult as jest.Mock).mockReturnValue([
          {
            order: 0,
            id: 123,
            search: {
              id: 123,
            },
            photo: {
              id: 123,
              file: "file",
              source: "source",
              dir: "dir",
            },
          },
          {
            order: 1,
            id: 456,
            search: {
              id: 123,
            },
            photo: {
              id: 456,
              file: "file",
              source: "source",
              dir: "dir",
            },
          },
        ] as EntitySearchResult[]);
      });

      test("should save the new search", async () => {
        await searchHelpers.doSearchWithCache(connection, {});
        expect(searchHelpers.saveNewSearch).toHaveBeenCalled();
      });

      test("should make the search in the library", async () => {
        await searchHelpers.doSearchWithCache(connection, {});
        expect(searchHelpers.doSearch).toHaveBeenCalled();
      });

      test("should save the search results", async () => {
        await searchHelpers.doSearchWithCache(connection, {});
        expect(searchHelpers.saveSearchResult).toHaveBeenCalled();
      });

      test("should return the results", async () => {
        expect(
          await searchHelpers.doSearchWithCache(connection, {})
        ).toMatchSnapshot();
      });

      test("should return the results with the pagination", async () => {
        expect(
          await searchHelpers.doSearchWithCache(connection, {}, 0, 1)
        ).toMatchSnapshot();
        expect(
          await searchHelpers.doSearchWithCache(connection, {}, 1, 1)
        ).toMatchSnapshot();
      });
    });

    describe("with saved results", () => {
      beforeEach(() => {
        (searchHelpers.findSavedSearch as jest.Mock).mockReturnValue({
          searchResults: [],
          filterBy: "",
          orderBy: "orderBy",
          source: "source",
          album: "album",
          id: 123,
        } as EntitySearch);

        (searchHelpers.fetchSearchResult as jest.Mock).mockReturnValue([
          {
            order: 0,
            id: 123,
            search: {
              id: 123,
            },
            photo: {
              id: 123,
              file: "file",
              source: "source",
              dir: "dir",
            },
          },
        ] as EntitySearchResult[]);
      });

      test("should NOT save a new search or do a new search", async () => {
        await searchHelpers.doSearchWithCache(connection, {});
        expect(searchHelpers.saveNewSearch).not.toHaveBeenCalled();
        expect(searchHelpers.doSearch).not.toHaveBeenCalled();
        expect(searchHelpers.saveSearchResult).not.toHaveBeenCalled();
      });

      test("should fetch the results from the saved search", async () => {
        await searchHelpers.doSearchWithCache(connection, {});
        expect(searchHelpers.fetchSearchResult).toHaveBeenCalled();
      });

      test("should return the saved photos", async () => {
        expect(
          await searchHelpers.doSearchWithCache(connection, {})
        ).toMatchSnapshot();
      });

      test("should return photo according to the pagination", async () => {
        await searchHelpers.doSearchWithCache(connection, {}, 5, 10);
        expect(
          (searchHelpers.fetchSearchResult as jest.Mock).mock.calls[0]
        ).toMatchSnapshot();
      });
    });
  });
});
