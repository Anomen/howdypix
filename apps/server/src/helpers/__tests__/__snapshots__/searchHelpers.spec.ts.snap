// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`searchHelpers doSearch should make the right query 1`] = `
Array [
  Object {
    "data": "yay",
  },
]
`;

exports[`searchHelpers doSearch should make the right query 2`] = `
Array [
  Array [
    Object {
      "order": Object {
        "birthtime": "ASC",
      },
      "where": Object {
        "dir": "album",
        "source": "source",
      },
    },
  ],
]
`;

exports[`searchHelpers doSearchWithCache with not-saved results should return the results 1`] = `
Array [
  Object {
    "id": 123,
    "order": 0,
    "photo": Object {
      "dir": "dir",
      "file": "file",
      "id": 123,
      "source": "source",
    },
    "search": Object {
      "id": 123,
    },
  },
  Object {
    "id": 456,
    "order": 1,
    "photo": Object {
      "dir": "dir",
      "file": "file",
      "id": 456,
      "source": "source",
    },
    "search": Object {
      "id": 123,
    },
  },
]
`;

exports[`searchHelpers doSearchWithCache with not-saved results should return the results with the pagination 1`] = `
Array [
  Object {
    "id": 123,
    "order": 0,
    "photo": Object {
      "dir": "dir",
      "file": "file",
      "id": 123,
      "source": "source",
    },
    "search": Object {
      "id": 123,
    },
  },
]
`;

exports[`searchHelpers doSearchWithCache with not-saved results should return the results with the pagination 2`] = `
Array [
  Object {
    "id": 456,
    "order": 1,
    "photo": Object {
      "dir": "dir",
      "file": "file",
      "id": 456,
      "source": "source",
    },
    "search": Object {
      "id": 123,
    },
  },
]
`;

exports[`searchHelpers doSearchWithCache with saved results should return photo according to the pagination 1`] = `
Array [
  undefined,
  Object {
    "album": "album",
    "filterBy": "",
    "id": 123,
    "orderBy": "orderBy",
    "searchResults": Array [],
    "source": "source",
  },
  5,
  10,
]
`;

exports[`searchHelpers doSearchWithCache with saved results should return the saved photos 1`] = `
Array [
  Object {
    "id": 123,
    "order": 0,
    "photo": Object {
      "dir": "dir",
      "file": "file",
      "id": 123,
      "source": "source",
    },
    "search": Object {
      "id": 123,
    },
  },
]
`;

exports[`searchHelpers fetchSearchResult should make the right query 1`] = `
Array [
  Object {
    "data": "yay",
  },
]
`;

exports[`searchHelpers fetchSearchResult should make the right query 2`] = `
Array [
  Array [
    Object {
      "order": Object {
        "order": "ASC",
      },
      "relations": Array [
        "photo",
      ],
      "where": Object {
        "order": FindOperator {
          "_getSql": undefined,
          "_multipleParameters": true,
          "_objectLiteralParameters": undefined,
          "_type": "between",
          "_useParameter": true,
          "_value": Array [
            1,
            6,
          ],
        },
        "search": Object {
          "album": "album",
          "filterBy": "",
          "id": 890,
          "orderBy": "orderBy",
          "searchResults": Array [],
          "source": "source",
        },
      },
    },
  ],
]
`;

exports[`searchHelpers findSavedSearch should make the right query 1`] = `
Object {
  "data": "yay",
}
`;

exports[`searchHelpers findSavedSearch should make the right query 2`] = `
Array [
  Array [
    Object {
      "where": Object {
        "album": "album",
        "filterBy": null,
        "orderBy": "DATE_ASC",
        "source": "source",
      },
    },
  ],
]
`;

exports[`searchHelpers saveNewSearch should make the right query 1`] = `
Object {
  "data": "yay",
}
`;

exports[`searchHelpers saveNewSearch should make the right query 2`] = `
Array [
  Array [
    Search {
      "album": "album",
      "filterBy": undefined,
      "id": undefined,
      "orderBy": "DATE_ASC",
      "searchResults": undefined,
      "source": "source",
    },
  ],
]
`;

exports[`searchHelpers saveSearchResult should make the right query 1`] = `
Array [
  Array [
    Object {
      "data": "yay",
    },
  ],
  Array [
    Object {
      "data": "yay",
    },
  ],
]
`;

exports[`searchHelpers saveSearchResult should make the right query 2`] = `
Array [
  Array [
    SearchResult {
      "id": undefined,
      "order": 0,
      "photo": Object {
        "id": 123,
      },
      "photoId": undefined,
      "search": Object {
        "id": 890,
      },
      "searchId": undefined,
    },
  ],
  Array [
    SearchResult {
      "id": undefined,
      "order": 1,
      "photo": Object {
        "id": 456,
      },
      "photoId": undefined,
      "search": Object {
        "id": 890,
      },
      "searchId": undefined,
    },
  ],
]
`;
